<a href="https://colab.research.google.com/github/shriraoshiwani11/FYND/blob/main/Python%20Task.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Importing the libraries

In [25]:
import numpy as np
import pandas as pd
import gdown
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')


### Downloading the dataset

In [26]:
# URL of the file on Google Drive
url = 'https://drive.google.com/uc?id=1PU3FYpjmHKvbOCxrFJZPYIJL8tYMdVpq'
output = 'DA.csv'

# Download the file
gdown.download(url, output, quiet=False)

# Read the CSV file
df = pd.read_csv(output)

Downloading...
From: https://drive.google.com/uc?id=1PU3FYpjmHKvbOCxrFJZPYIJL8tYMdVpq
To: /content/DA.csv
100%|██████████| 25.5M/25.5M [00:01<00:00, 17.0MB/s]


## 1) Importing the dataset and usual Analysis

In [27]:
df.head(10)

date ordering_channel      mode_of_payment    bag_id  \
0  2021-04-24 22:49:19 UTC      MARKETPLACE  FY_FLIPKART_ASSURED  10299882   
1  2021-08-11 20:43:46 UTC            ECOMM                ECOMM  11609617   
2  2021-03-09 17:31:07 UTC      MARKETPLACE  FY_FLIPKART_ASSURED   9907918   
3  2021-09-29 22:57:12 UTC      MARKETPLACE  FY_FLIPKART_ASSURED  12331936   
4  2021-09-29 15:30:17 UTC      MARKETPLACE            FY_AMAZON  12326311   
5  2021-03-14 23:01:20 UTC      MARKETPLACE            FY_AMAZON   9955822   
6  2021-08-15 17:42:23 UTC            ECOMM                ECOMM  11671957   
7  2021-03-09 15:05:45 UTC      MARKETPLACE            FY_AMAZON   9906558   
8  2021-10-25 20:01:50 UTC      MARKETPLACE            FY_AMAZON  12863928   
9  2021-09-18 09:46:46 UTC      MARKETPLACE            FY_MYNTRA  12172254   

            shipment_id                      affiliate_bag_id  \
0  16194176947811624583  dbd766d8-dfe2-49d6-8dea-3808c2ff1998   
1  16292033335741984031                              11609617   
2  16152912668821225728  842acd6b-f3ac-4662-b92d-31d177b25b05   
3  16329364323991697144  1bde7407-e1b1-41ba-9f90-bb9c1feec7ed   
4  16329096170171681635                              12326311   
5  16157896956771977367                             DVJqQdrxg   
6  16290295437581443455                              11671957   
7  16152831492231474899                             DcZWDYvj8   
8  16351723095571457749                              12863928   
9  16319393992991935120                            6524924285   

               order_id      Itemcode       sku_code           ean  ...  \
0  FYMP6084531601CC91EE   S9C5089-040  8907181811994           NaN  ...   
1  FY6113E92A014E6066CA    927389-036   886549834401           NaN  ...   
2  FYMP604763820183E9EF  S1810401-010  8907181713120           NaN  ...   
3  FYMP6154A1F00169BE96    AV5014-010  8907181795836           NaN  ...   
4  FYMP6154393001A8F48D        FJ9217  4060512773966           NaN  ...   
5  FYMP604E486801EA37D0    AR4840-222  8907181468365           NaN  ...   
6  FY611904A701E00B4B76   S8L4893-414  8907181711256           NaN  ...   
7  FYMP6047417001ABE5D6   S8C4916-949            NaN  8.907182e+12  ...   
8  FYMP6176BFD50137461E        FL5726  4062051160528           NaN  ...   
9  FYMP6145682D011BDD65    215489-1CL            NaN  8.907182e+12  ...   

      Store_City    Store_State             del_cust_name del_cust_mobile  \
0       DEHRADUN    UTTARAKHAND             Saquib Younus      9748408797   
1        MATHURA  UTTAR PRADESH            Rahul Tamaskar      9685761206   
2           PUNE    MAHARASHTRA              Gaurav Kumar      8700697973   
3      NEW DELHI          DELHI               Sandip Jena      9348036053   
4        MATHURA  UTTAR PRADESH   Amazon default customer       234242340   
5           PUNE    MAHARASHTRA               Raman jain       7223057217   
6       DURGAPUR    WEST BENGAL             Madhav Padwal      8451080005   
7           PUNE    MAHARASHTRA  VENUGOPALARAO CHELIKANI       9848437337   
8           KOTA      RAJASTHAN   Amazon default customer       234242340   
9  GREATER NOIDA  UTTAR PRADESH            Varsha Ghute        9999999999   

    l1_category          l3_category                     Company_Name  \
0  ["Footwear"]           Flip Flops  SSIPL LIFESTYLE PRIVATE LIMITED   
1  ["Clothing"]   Activewear Joggers  SSIPL LIFESTYLE PRIVATE LIMITED   
2  ["Clothing"]   Activewear Joggers  SSIPL LIFESTYLE PRIVATE LIMITED   
3  ["Footwear"]             Sneakers  SSIPL LIFESTYLE PRIVATE LIMITED   
4  ["Clothing"]             T-Shirts  SSIPL LIFESTYLE PRIVATE LIMITED   
5  ["Footwear"]  Indoor Sports Shoes  SSIPL LIFESTYLE PRIVATE LIMITED   
6  ["Footwear"]  Indoor Sports Shoes  SSIPL LIFESTYLE PRIVATE LIMITED   
7  ["Footwear"]           Flip Flops  SSIPL LIFESTYLE PRIVATE LIMITED   
8  ["Clothing"]               Shorts  SSIPL LIFESTYLE PRIVATE LIMITED   
9  ["Clothing"]               Capris  SSIPL

In [22]:
#Shape of the Dataset
df.shape

(46436, 42)

In [23]:
#List of all the columns
df.columns

Index(['date', 'ordering_channel', 'mode_of_payment', 'bag_id', 'shipment_id',
       'affiliate_bag_id', 'order_id', 'Itemcode', 'sku_code', 'ean', 'upc',
       'item_gender', 'article_qnty', 'ProdName', 'Size', 'Brand', 'Gender',
       'payment_mode', 'ESP', 'Amount_Paid', 'unit_price', 'Cur_Status',
       'Address_1', 'Pincode', 'Delivery_State', 'Location',
       'Marketplace_awb_number', 'Cust_Name', 'Mobile', 'Store_Code',
       'Cur_Store_Code', 'Store_Address', 'Store_City', 'Store_State',
       'del_cust_name', 'del_cust_mobile', 'l1_category', 'l3_category',
       'Company_Name', 'hsn_code', 'invoice_id', 'MRP'],
      dtype='object')

In [28]:
#Basic details of all the columns
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46436 entries, 0 to 46435
Data columns (total 42 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   date                    46436 non-null  object 
 1   ordering_channel        46436 non-null  object 
 2   mode_of_payment         46436 non-null  object 
 3   bag_id                  46436 non-null  int64  
 4   shipment_id             46436 non-null  uint64 
 5   affiliate_bag_id        46436 non-null  object 
 6   order_id                46436 non-null  object 
 7   Itemcode                46436 non-null  object 
 8   sku_code                30083 non-null  object 
 9   ean                     16353 non-null  float64
 10  upc                     0 non-null      float64
 11  item_gender             35654 non-null  object 
 12  article_qnty            46436 non-null  int64  
 13  ProdName                46436 non-null  object 
 14  Size                    45352 non-null

In [29]:
# Statistical summary of data
df.describe()

bag_id   shipment_id           ean  upc  article_qnty  Gender  \
count  4.643600e+04  4.643600e+04  1.635300e+04  0.0  46436.000000     0.0   
mean   1.135590e+07  1.626905e+19  8.733145e+12  NaN      5.070527     NaN   
std    7.925085e+05  5.157953e+16  9.537656e+11  NaN     33.783096     NaN   
min    9.841155e+06  1.614624e+19  1.918842e+11  NaN      0.000000     NaN   
25%    1.075046e+07  1.623841e+19  8.907182e+12  NaN      2.000000     NaN   
50%    1.125842e+07  1.626703e+19  8.907182e+12  NaN      2.000000     NaN   
75%    1.203552e+07  1.631039e+19  8.907182e+12  NaN      4.000000     NaN   
max    1.287622e+07  1.635242e+19  8.907182e+12  NaN   1480.000000     NaN   

                ESP   Amount_Paid    unit_price        Pincode      hsn_code  \
count  46436.000000  46436.000000  46436.000000   46436.000000  4.643600e+04   
mean    1134.788754   1136.579882   1133.918341  433982.232126  6.263286e+07   
std     1203.287900   1201.727176   1203.134237  227987.106673  5.743718e+06   
min        0.000000      0.000000      0.000000  110001.000000  3.302901e+07   
25%      499.000000    508.000000    499.000000  221002.000000  6.402999e+07   
50%      799.000000    799.000000    799.000000  411042.500000  6.404119e+07   
75%      999.000000   1008.299988    999.000000  600114.000000  6.404119e+07   
max    18999.000000  18999.000000  18999.000000  855116.000000  9.506999e+07   

                MRP  
count  46436.000000  
mean    2146.942534  
std     1413.185562  
min        0.000000  
25%      999.000000  
50%     1999.000000  
75%     2999.000000  
max    18999.000000

## Checking for null values

In [30]:
df.isna().sum()

date                          0
ordering_channel              0
mode_of_payment               0
bag_id                        0
shipment_id                   0
affiliate_bag_id              0
order_id                      0
Itemcode                      0
sku_code                  16353
ean                       30083
upc                       46436
item_gender               10782
article_qnty                  0
ProdName                      0
Size                       1084
Brand                         0
Gender                    46436
payment_mode                 15
ESP                           0
Amount_Paid                   0
unit_price                    0
Cur_Status                    0
Address_1                    24
Pincode                       0
Delivery_State                1
Location                      0
Marketplace_awb_number    10275
Cust_Name                   901
Mobile                      901
Store_Code                    0
Cur_Store_Code                0
Store_Address                 0
Store_City                    0
Store_State                   0
del_cust_name                 0
del_cust_mobile               0
l1_category                   0
l3_category                   0
Company_Name                  0
hsn_code                      0
invoice_id                 9478
MRP                           0
dtype: int64

From above result, the missing values are present in the following columns.


1. sku_code	16353
2. ean	30083
3. upc	46436
4. item_gender 10782
5. Size	1084
6. Gender	46436
7. payment_mode	15
8. Address_1	24
9. Delivery_State	1
10. Marketplace_awb_number	10275
11. Cust_Name	901
12. Mobile	901
13. invoice_id	9478




##Filling the Null values with default values

In [55]:
#Default value for numerical column = 0
#Default value for string column = "missing"

default_values = {
    'sku_code': "missing",
    'ean': "missing",
    'upc': "missing",
    'item_gender': "missing" ,
    'Size': "missing",
    'Gender': "missing",
    'payment_mode': "missing",
    'Address_1': "missing" ,
    'Delivery_State': "missing",
    'Marketplace_awb_number': "missing",
    'Cust_Name': "missing",
    'Mobile': "missing" ,
    'invoice_id': "missing"
}

In [56]:
df2 = df.fillna(default_values)

In [57]:
#After filling the null values with default values
df2.isna().sum()

date                      0
ordering_channel          0
mode_of_payment           0
bag_id                    0
shipment_id               0
affiliate_bag_id          0
order_id                  0
Itemcode                  0
sku_code                  0
ean                       0
upc                       0
item_gender               0
article_qnty              0
ProdName                  0
Size                      0
Brand                     0
Gender                    0
payment_mode              0
ESP                       0
Amount_Paid               0
unit_price                0
Cur_Status                0
Address_1                 0
Pincode                   0
Delivery_State            0
Location                  0
Marketplace_awb_number    0
Cust_Name                 0
Mobile                    0
Store_Code                0
Cur_Store_Code            0
Store_Address             0
Store_City                0
Store_State               0
del_cust_name             0
del_cust_mobile           0
l1_category               0
l3_category               0
Company_Name              0
hsn_code                  0
invoice_id                0
MRP                       0
dtype: int64

##Correcting the data types

In [58]:
#Defining the data types
target_types = {
    'date': 'datetime64',
    'ordering_channel': 'category',
    'mode_of_payment': 'category',
    'bag_id': 'str',
    'shipment_id': 'str',
    'affiliate_bag_id': 'str',
    'order_id': 'str',
    'Itemcode': 'str',
    'sku_code': 'str',
    'ean': 'str',
    'upc': 'str',
    'item_gender': 'category',
    'article_qnty': 'int',
    'ProdName': 'str',
    'Size': 'str',
    'Brand': 'str',
    'Gender': 'category',
    'payment_mode': 'category',
    'ESP': 'float',
    'Amount_Paid': 'float',
    'unit_price': 'float',
    'Cur_Status': 'category',
    'Address_1': 'str',
    'Pincode': 'int',
    'Delivery_State': 'str',
    'Location': 'str',
    'Marketplace_awb_number': 'str',
    'Cust_Name': 'str',
    'Mobile': 'str',
    'Store_Code': 'str',
    'Cur_Store_Code': 'str',
    'Store_Address': 'str',
    'Store_City': 'str',
    'Store_State': 'str',
    'del_cust_name': 'str',
    'del_cust_mobile': 'str',
    'l1_category': 'str',
    'l3_category': 'str',
    'Company_Name': 'str',
    'hsn_code': 'str',
    'invoice_id': 'str',
    'MRP': 'float'
}

# Convert data types

df2[column] = df2[column].astype(dtype)

print(df2.dtypes)

date                       object
ordering_channel           object
mode_of_payment            object
bag_id                      int64
shipment_id                uint64
affiliate_bag_id           object
order_id                   object
Itemcode                   object
sku_code                   object
ean                        object
upc                        object
item_gender                object
article_qnty                int64
ProdName                   object
Size                       object
Brand                      object
Gender                     object
payment_mode               object
ESP                       float64
Amount_Paid               float64
unit_price                float64
Cur_Status                 object
Address_1                  object
Pincode                     int64
Delivery_State             object
Location                   object
Marketplace_awb_number     object
Cust_Name                  object
Mobile                     object
Store_Code    

In [62]:
#Saving the cleaned data set
df2.to_csv('cleaned_dataset.csv', index=False)

In [61]:
df2.head()

date ordering_channel      mode_of_payment    bag_id  \
0  2021-04-24 22:49:19 UTC      MARKETPLACE  FY_FLIPKART_ASSURED  10299882   
1  2021-08-11 20:43:46 UTC            ECOMM                ECOMM  11609617   
2  2021-03-09 17:31:07 UTC      MARKETPLACE  FY_FLIPKART_ASSURED   9907918   
3  2021-09-29 22:57:12 UTC      MARKETPLACE  FY_FLIPKART_ASSURED  12331936   
4  2021-09-29 15:30:17 UTC      MARKETPLACE            FY_AMAZON  12326311   

            shipment_id                      affiliate_bag_id  \
0  16194176947811624583  dbd766d8-dfe2-49d6-8dea-3808c2ff1998   
1  16292033335741984031                              11609617   
2  16152912668821225728  842acd6b-f3ac-4662-b92d-31d177b25b05   
3  16329364323991697144  1bde7407-e1b1-41ba-9f90-bb9c1feec7ed   
4  16329096170171681635                              12326311   

               order_id      Itemcode       sku_code      ean  ... Store_City  \
0  FYMP6084531601CC91EE   S9C5089-040  8907181811994  missing  ...   DEHRADUN   
1  FY6113E92A014E6066CA    927389-036   886549834401  missing  ...    MATHURA   
2  FYMP604763820183E9EF  S1810401-010  8907181713120  missing  ...       PUNE   
3  FYMP6154A1F00169BE96    AV5014-010  8907181795836  missing  ...  NEW DELHI   
4  FYMP6154393001A8F48D        FJ9217  4060512773966  missing  ...    MATHURA   

     Store_State            del_cust_name del_cust_mobile   l1_category  \
0    UTTARAKHAND            Saquib Younus      9748408797  ["Footwear"]   
1  UTTAR PRADESH           Rahul Tamaskar      9685761206  ["Clothing"]   
2    MAHARASHTRA             Gaurav Kumar      8700697973  ["Clothing"]   
3          DELHI              Sandip Jena      9348036053  ["Footwear"]   
4  UTTAR PRADESH  Amazon default customer       234242340  ["Clothing"]   

          l3_category                     Company_Name  hsn_code   invoice_id  \
0          Flip Flops  SSIPL LIFESTYLE PRIVATE LIMITED  64041190      missing   
1  Activewear Joggers  SSIPL LIFESTYLE PRIVATE LIMITED  61034300  SS030/00535   
2  Activewear Joggers  SSIPL LIFESTYLE PRIVATE LIMITED  61103010  T0090/12371   
3            Sneakers  SSIPL LIFESTYLE PRIVATE LIMITED  64041190      missing   
4            T-Shirts  SSIPL LIFESTYLE PRIVATE LIMITED  61099090  ss030/00906   

      MRP  
0   849.0  
1  2195.0  
2  1199.0  
3  1999.0  
4   999.0  

[5 rows x 42 columns]